### PR TITLE
Separate box blur into two kernels.

### PR DESCRIPTION
**Pull request by jonawals**
# What Is This Optimization?

This PR optimizes the blur pass of the Editor Mode Feedback effect for Focus Mode to reduce the number of texture lookups required to convolve the input color buffer with the box blur kernel.

# Why Is This Optimization Needed?

The current implementation uses a 2d convolution kernel for a time complexity of _n<sup>2</sup>_ per texel in the input color buffer, where _n_ is the width of the kernel. This PR implements a common optimization where the time complexity is reduced to _n_.

# How Is This Optimization Implemented?

The blur pass has been separated out into discrete horizontal and vertical passes. Each pass convolves the input color buffer with a 1d kernel either horizontaly or vertically, with the output of whichever pass is first being the input of whichever pass is second (i.e, input color buffer -> horizontal pass -> vertical pass ->output color buffer).

# What Are The Results Of This Optimization?

### Original Approach

Pass | Duration (ms)
-- | --
BlurPass | 0.709

### Proposed Approach

Pass | Duration (ms)
-- | --
BlurHPass | 0.1
BlurVPass | 0.096
**Total** | **0.196**

### Discussion

This PR delivers a 351% performance increase for the blur pass of the Editor Mode Feedback System.

# Testing Done

Manual testing was performed in the editor by comparing the performance and output of the original and optimized blur algorithms.

Signed-off-by: John <jonawals@amazon.com>

### CODE REVIEW

File: Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 // Temporary measure for setting the blur pass shader parameters at runtime until GHI 3455 is implemented
 AZ_EDITOR_MODE_PASS_TRANSITION_CVARS(cl_editorModeBlurPass, 0.0f, 0.0f, 20.0f, 1.0f);
-AZ_EDITOR_MODE_PASS_CVAR(float, cl_editorModeBlurPass, KernalWidth, 5.0f);
```
Comment (moraaar) [inline Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp:18]: In the PR description says `Kernel` and in here it's `Kernal`, what's the difference?

File: Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp [hunk 2]
```diff
@@ -15,7 +15,7 @@
...
+AZ_EDITOR_MODE_PASS_CVAR(float, cl_editorModeBlurPass, KernalHalfWidth, 5.0f);
```
Comment (moraaar) [inline Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp:18]: nit: Worth mentioning the unit in the name just to be clear? maybe `KernalHalfWidthTexels`.

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl [hunk 3]
```diff
@@ -12,27 +12,24 @@
...
-        for(float v = -kernalWidth; v <= kernalWidth; v+=1.)
-        {
-            weights += 1.;
-            outColor += PassSrg::m_framebuffer.Sample(PassSrg::PointSampler, (unormalizedUv + float2(u,v)) * PassSrg::m_maskDimensions.zw).rgb;
-        }
+        outColor += PassSrg::m_framebuffer.Sample(PassSrg::PointSampler, (unormalizedUv + (float2(i,i) * uvDir)) * PassSrg::m_maskDimensions.zw).rgb;
     }
 
-    return outColor / weights;
+    return outColor / (kernalHalfWidth * 2 + 1);
```
Comment (moraaar) [inline Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl:32]: To be consistent, and potentially avoid different compilers doing different things, better to match the type of the constants to `kernalHalfWidth` type by using 2.0 and 1.0.

**Review by moraaar — APPROVED** [review]
Great performance gain!

File: Gems/Atom/Feature/Common/Assets/Passes/EditorModeBlurHorizontal.pass [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                            // Horizontal blur
+                            "Name": "m_direction",
+                            "Value": 0
+                        }
+                    ]
+                }
+            }
+        }
+    }
+}
```
Comment (hultonha) [inline Gems/Atom/Feature/Common/Assets/Passes/EditorModeBlurHorizontal.pass:87]: nit: Missing EOL character (you can set VSCode to always add one which is a really useful option to set)

File: Gems/Atom/Feature/Common/Assets/Passes/EditorModeBlurParent.pass [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                                "Pass": "BlurHorizontalPass",
+                                "Attachment": "OutputColor"
+                            }
+                        }
+                    ]
+                }
+            ]
+        }
+    }
+}
```
Comment (hultonha) [inline Gems/Atom/Feature/Common/Assets/Passes/EditorModeBlurParent.pass:100]: nit: Repeated (see above)

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl [hunk 6]
```diff
@@ -43,8 +40,9 @@ PSOutput MainPS(VSOutput IN)
     const float mask = PassSrg::m_entityMask.Sample(PassSrg::PointSampler, IN.m_texCoord).r;
     
     // Blur effect
-    float3 finalEffect = Blur(IN.m_texCoord, PassSrg::m_kernalWidth);
-  
+    float2 blurDir = (PassSrg::m_direction == 0) ? float2(1.0, 0.0) : float2(0.0, 1.0);
```
Comment (hultonha) [inline Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl:43]: nit: Prefer `blurDirection` as opposed to `blurDir` (this type of naming is mentioned in the coding standards)

File: Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp (hunk 1 above)
Comment (jonawals) [reply Gems/Atom/Feature/Common/Code/Source/PostProcessing/EditorModeBlurPass.cpp:18]: "Kernel" spelling is more consistent with the rest of the code base so i will change all instances of "kernal" to that 👍

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl (hunk 3 above)
Comment (jonawals) [reply Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl:32]: Fixed 👍

File: Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl [hunk 7]
```diff
@@ -12,27 +12,24 @@
 partial ShaderResourceGroup PassSrg : SRG_PerPass_WithFallback
 {
     //! Width of kernal to apply box blur effect.
-    float m_kernalWidth;
+    float m_kernalHalfWidth;
```
[the remark this answers is not in the document]
Comment (jonawals) [reply Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/EditorModeBlur.azsl:15]: Fixed 👍

**Review by VickyAtAZ — COMMENTED** [review]
It seems the different between the BlurVertical and BlurHorizontal is the ShaderDataMappings. 
One simplification is to move the passData in parentPass's pass request, and you only have one EditorModeBlur.pass template. The PassData section in EditorModeBlur.pass can be removed.

In parent pass the pass requests to create two passed may look like this
```
"PassRequests": [
                {
                    "Name": "BlurHorizontalPass",
                    "TemplateName": "EditorModeBlurTemplate",
                    "Connections": [
					...
                    ],
					"PassData": {
						"$type": "FullscreenTrianglePassData",
						"ShaderAsset": {
							"FilePath": "Shaders/PostProcessing/EditorModeBlur.shader"
						},
						"PipelineViewTag": "MainCamera",
						"ShaderDataMappings": {
							"UintMappings": [
								{
									// Horizontal blur
									"Name": "m_direction",
									"Value": 0
								}
							]
						}
					}
                },
                {
                    "Name": "BlurVerticalPass",
                    "TemplateName": "EditorModeBlurTemplate",
                    "Connections": [
					...
                    ],
					"PassData": {
						"$type": "FullscreenTrianglePassData",
						"ShaderAsset": {
							"FilePath": "Shaders/PostProcessing/EditorModeBlur.shader"
						},
						"PipelineViewTag": "MainCamera",
						"ShaderDataMappings": {
							"UintMappings": [
								{
									// Vertical blur
									"Name": "m_direction",
									"Value": 1
								}
							]
						}
					}
                }
            ]
```

**Review by VickyAtAZ — APPROVED** [review]

**Comment by jonawals** [comment]
@AMZN-alexpete That was my first thought as well (and there’s certainly an argument to be made for a general library of common effects and utils) but as the blurring itself is only a few lines of code in the shader it would be a non-trivial job to wrap it up as a generic algorithm and then attempt to retrofit all other blur usages to use it.

**Comment by jonawals** [comment]
I will be merging this branch into the parent branch so that the parent can then be merged into dev. If anyone has any further comments then reach out to me and we can follow up on them in a separate PR.

**Pull request merged**